IPython/core/oinspect.py

```diff
@@ -821,8 +822,22 @@ def _info(self, obj, oname='', info=None, detail_level=0) -> dict:
                 pass
 
         # Add docstring only if no source is to be shown (avoid repetitions).
-        if ds and out.get('source', None) is None:
-            out['docstring'] = ds
+        if ds:
+            skip_ds = False
+            if out.get('source'):
+                try:
+                    def_node, = ast.parse(dedent(out['source'])).body
+                    node_docstring = ast.get_docstring(def_node)
+                except Exception:
+                    # The source can become invalid (because it is re-fetched
+                    # from the source file) so the parse can fail in arbitrary
+                    # ways.
+                    pass
+                else:
+                    if node_docstring == ds:
+                        skip_ds = True
```

Comment:
Contents of this `if` could be a helper function `skip_ds = docstring_matches_source(out)`

Reply:
A helper for an equality check seems a bit overkill, no?

Reply:
I mean the entire try/catch

Reply:
I think the whole method needs some refactoring, in particular so that we can compare with the *actual* docstring and not some of the autogenerated fill-in docstrings that come from the top of the method...

Reply:
All the more reason to wrap this into a function, so that you can easily swap out how the docstring is obtained later

Reply:
done